File: 61. Rotate List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == NULL) return head;
        ListNode* p = head;
        int length = 0;
        while (p != NULL && p->next != NULL)
        {
            length ++;
            p = p->next;
        }
        length ++;
        p->next = head;
        k %= length;
        for (int i = 0; i < length - k; i++)
            p = p->next;
        ListNode* newhead = p->next;
        p->next = NULL;
        return newhead;
    }
};

File: 61. Rotate List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if not head:
            return head
        p = head
        length = 0
        while p and p.next:
            length += 1
            p = p.next
        length += 1
        k = k % length
        p.next = head
        for i in xrange(length - k):
            p = p.next
        newhead = p.next
        p.next = None
        return newhead
